crates/sound/src/stream.rs

```diff
@@ -0,0 +1,305 @@
+///   The driver transfers data to/from the stream.
+///
+/// - `STOP`
+///
+///   The device stops the stream (mute, putting into non-running state, etc).
+///
+///   Possible valid transitions: `START`, `RELEASE`.
+///
+/// - `RELEASE`
+///
+///   The device releases the stream (frees resources, etc).
+///
+///   Possible valid transitions: `SET PARAMETERS`, `PREPARE`.
+///
+/// ```text
+/// +---------------+ +---------+ +---------+ +-------+ +-------+
+/// | SetParameters | | Prepare | | Release | | Start | | Stop  |
+/// +---------------+ +---------+ +---------+ +-------+ +-------+
+///         |              |           |          |         |
+///         |-             |           |          |         |
+///         ||             |           |          |         |
+///         |<             |           |          |         |
+///         |              |           |          |         |
+///         |------------->|           |          |         |
+///         |              |           |          |         |
+///         |<-------------|           |          |         |
+///         |              |           |          |         |
+///         |              |-          |          |         |
+///         |              ||          |          |         |
+///         |              |<          |          |         |
+///         |              |           |          |         |
+///         |              |--------------------->|         |
+///         |              |           |          |         |
+///         |              |---------->|          |         |
+///         |              |           |          |         |
+///         |              |           |          |-------->|
+///         |              |           |          |         |
+///         |              |           |          |<--------|
+///         |              |           |          |         |
+///         |              |           |<-------------------|
+///         |              |           |          |         |
+///         |<-------------------------|          |         |
+///         |              |           |          |         |
+///         |              |<----------|          |         |
+/// ```
+#[derive(Debug, Default, Copy, Clone)]
+pub enum PCMState {
+    #[default]
+    #[doc(alias = "VIRTIO_SND_R_PCM_SET_PARAMS")]
+    SetParameters,
+    #[doc(alias = "VIRTIO_SND_R_PCM_PREPARE")]
+    Prepare,
+    #[doc(alias = "VIRTIO_SND_R_PCM_RELEASE")]
+    Release,
+    #[doc(alias = "VIRTIO_SND_R_PCM_START")]
+    Start,
+    #[doc(alias = "VIRTIO_SND_R_PCM_STOP")]
+    Stop,
+}
+
+macro_rules! set_new_state {
+    ($new_state_fn:ident, $new_state:expr, $($valid_source_states:tt)*) => {
+        pub fn $new_state_fn(&mut self) -> Result<()> {
+            if !matches!(self, $($valid_source_states)*) {
+                return Err(Error::InvalidStateTransition(*self, $new_state));
+            }
+            *self = $new_state;
```

Comment:
If I understand correctly, here defines a macro to check if a cmd is allowed in set of potential current states. For example, the `prepare` cmd is only allowed if current state is `prepare`, `set_parameters` and `release`. If the cmd is executed in a different current state, the function fails. If the cmd is allowed based on the virtio spec, the state transits to the triggered command. This verification happens in the audio_backend itself (see https://github.com/epilys/vhost-device/blob/904391c35b83e885d6d36be48ccd9f72742785c9/crates/sound/src/audio_backends/alsa.rs#L432). In our case, we are doing it in the device before calling any `audio_backend` function. This is the purpose of the `get_stream_state()` method, i.e., to get the stream state and then verify that the cmd is allowed. AFAIU, the virtio spec does not tell what is the destination state during a transiton but only the valid current state in which a cmd is allowed to be executed. The outgoing state may vary from backend to backend so the PCM state-machine for alsa may not be the same than the pipewire PCM-state machine although both conform to the virtio spec.

Reply:
The state you change to is what "Possible valid transitions" in the spec means though, right?

Reply:
I don't understand the point you're trying to make though. The code here checks the valid transitions, and yes what the transition will be is the caller's decision.

Reply:
AIUI all this ensures is the backend doesn't attempt to transition to a state the spec doesn't allow.

Reply:
@stsquad that is what the code this comment points to does, yes

Reply:
Thanks both for the clarification. This is a minor comment that we can discuss further at a later time. To illustrate my point, in the pipewire backend, we decided to stay in the `Self::Prepare` state if the `set_parameter` cmd is issued and the current state is `Self::Prepare`. It was simpler to do it like this. I think the PCM state machine shall be audio-backend-dependent. I may be wrong but I think the spec does not tell you what is the next state but only the cmds that are valid in a given current state. I will try to clarify this with the authors of the spec.

Reply:
This code allows that behavior (staying from Prepare to Prepare), so **_I am still not understanding_** what your point is about. I think you have probably not understood yourself.